File: system/ulib/zx/process.cpp
// Copyright 2016 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

#include <lib/zx/process.h>

#include <zircon/syscalls.h>

#include <lib/zx/job.h>
#include <lib/zx/thread.h>
#include <lib/zx/vmar.h>

namespace zx {

zx_status_t process::create(const job& job, const char* name, uint32_t name_len, uint32_t flags,
                            process* proc, vmar* vmar) {
    zx_handle_t proc_h;
    zx_handle_t vmar_h;
    zx_status_t status = zx_process_create(job.get(), name, name_len, flags, &proc_h, &vmar_h);
    if (status < 0) {
        proc->reset(ZX_HANDLE_INVALID);
        vmar->reset(ZX_HANDLE_INVALID);
    } else {
        proc->reset(proc_h);
        vmar->reset(vmar_h);
    }
    return status;
}

zx_status_t process::start(const thread& thread_handle, uintptr_t entry,
                           uintptr_t stack, handle arg_handle,
                           uintptr_t arg2) const {
    zx_handle_t arg_h = arg_handle.release();
    zx_status_t result =
        zx_process_start(get(), thread_handle.get(), entry, stack, arg_h, arg2);
    if (result < 0)
        zx_handle_close(arg_h);
    return result;
}

} // namespace zx

File: garnet/public/rust/fuchsia-zircon/src/process.rs
// Copyright 2017 The Fuchsia Authors. All rights reserved.
// Use of this source code is governed by a BSD-style license that can be
// found in the LICENSE file.

//! Type-safe bindings for Zircon processes.

use crate::{AsHandleRef, HandleBased, Handle, HandleRef, Status, Thread};
use crate::ok;

use fuchsia_zircon_sys as sys;

/// An object representing a Zircon process.
///
/// As essentially a subtype of `Handle`, it can be freely interconverted.
#[derive(Debug, Eq, PartialEq, Ord, PartialOrd, Hash)]
#[repr(transparent)]
pub struct Process(Handle);
impl_handle_based!(Process);

impl Process {
    /// Similar to `Thread::start`, but is used to start the first thread in a process.
    ///
    /// Wraps the
    /// [zx_process_start](https://fuchsia.googlesource.com/fuchsia/+/master/zircon/docs/syscalls/process_start.md)
    /// syscall.
    pub fn start(
        &self,
        thread: &Thread,
        entry: usize,
        stack: usize,
        arg1: Handle,
        arg2: usize,
    ) -> Result<(), Status> {
        let process_raw = self.raw_handle();
        let thread_raw = thread.raw_handle();
        let arg1 = arg1.into_raw();
        ok(unsafe {
            sys::zx_process_start(
                process_raw, thread_raw, entry, stack, arg1, arg2)
        })
    }

    /// Create a thread inside the current process.
    ///
    /// Wraps the
    /// [zx_thread_create](https://fuchsia.googlesource.com/fuchsia/+/master/zircon/docs/syscalls/thread_create.md)
    /// syscall.
    pub fn create_thread(&self, name: &[u8]) -> Result<Thread, Status> {
        let process_raw = self.raw_handle();
        let name_ptr = name.as_ptr();
        let name_len = name.len();
        let options = 0;
        let mut thread_out = 0;
        let status = unsafe { sys::zx_thread_create(
            process_raw,
            name_ptr,
            name_len,
            options,
            &mut thread_out,
        ) };
        ok(status)?;
        unsafe {
            Ok(Thread::from(Handle::from_raw(thread_out)))
        }
    }
}
